src/main/scala/com/microsoft/hyperspace/actions/VacuumOutdatedAction.scala

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.actions
+
+import org.apache.hadoop.fs.Path
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.actions.Constants.States.{ACTIVE, VACUUMINGOUTDATED}
+import com.microsoft.hyperspace.index.{IndexConstants, IndexDataManager, IndexLogEntry, IndexLogManager, LogEntry}
+import com.microsoft.hyperspace.telemetry.{AppInfo, HyperspaceEvent, VacuumOutdatedActionEvent}
+import com.microsoft.hyperspace.util.FileUtils
+
+/**
+ * Vacuum outdated data of indexes.
+ *
+ * Algorithm:
+ *  - Delete every version except the latest versions.
+ */
+class VacuumOutdatedAction(
+    final override protected val logManager: IndexLogManager,
+    dataManager: IndexDataManager)
+    extends Action {
+  override def logEntry: LogEntry = {
+    logManager.getLog(baseId).getOrElse {
+      throw HyperspaceException("LogEntry must exist for vacuum outdated data operation")
+    }
+  }
+
+  override def transientState: String = VACUUMINGOUTDATED
+
+  override def finalState: String = ACTIVE
+
+  override def validate(): Unit = {
+    if (!logEntry.state.equalsIgnoreCase(ACTIVE)) {
+      throw HyperspaceException(
+        s"VacuumOutdated is only supported in $ACTIVE state. " +
+          s"Current state is ${logEntry.state}.")
+    }
+  }
+
+  final override def op(): Unit = {
+    // Delete unused directory first, then delete unused files in used directories.
+    val indexVersionsInUse: Set[Int] = logEntry match {
+      case indexLogEntry: IndexLogEntry =>
+        dataVersionInfos(indexLogEntry)
+
+      case other =>
+        throw HyperspaceException(
+          s"VacuumOutdated is not supported for log entry class ${other.getClass.getName}")
+    }
+
+    // Delete version directories not used.
+    dataManager.getAllVersionIds().foreach { id =>
+      if (!indexVersionsInUse.contains(id)) {
+        dataManager.delete(id)
+      }
+    }
+
+    val filesInUse = logEntry match {
+      case indexLogEntry: IndexLogEntry =>
+        indexLogEntry.content.fileInfos.map { info =>
+          info.name
+        }
+    }
+
+    // Delete unused files
+    dataManager.getAllFilePaths().foreach { path =>
+      if (path.getName.startsWith("part-") && !filesInUse.contains(path.toString)) {
```

Comment:
All file paths will start with "part-" so I don't think the condition is necessary.

Reply:
Hmm. There was a file "_SUCCESS" to indicate success of tasks. Is it ok to delete it?